File: src/test/java/utils/ProjectPaths.java
package utils;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.net.URI;
import java.net.URISyntaxException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Objects;
import java.util.Optional;
import java.util.stream.Collectors;


public class ProjectPaths {
    private static final ProjectPaths INSTANCE = new ProjectPaths();

    private ProjectPaths() {
    }

    private static @NotNull String getPathName(final @NotNull Path path) {
        return Optional.ofNullable(path.getFileName()).map(Object::toString).orElse("");
    }

    public static @NotNull Path getProjectRoot() {
        return Objects.requireNonNull(INSTANCE._getProjectRoot());
    }

    private @Nullable Path _getProjectRoot() {
        var directory = getProjectRootByClass();
        while (directory != null) {
            try {
                try (final var fileList = Files.list(directory)) {
                    final var children = fileList.map(ProjectPaths::getPathName).collect(Collectors.toSet());
                    if (children.contains("gradle") && children.contains("src")) {
                        return directory;
                    }
                }
            } catch (final IOException ignore) {
            }
            directory = directory.getParent();
        }
        return null;
    }

    private @Nullable Path getProjectRootByClass() {
        final var replacedName = this.getClass().getName().replace(".", "/");
        final var url = getClass().getResource("/" + replacedName + ".class");
        if (url == null) {
            return null;
        }
        try {
            return switch (url.getProtocol()) {
                case "file" -> Paths.get(url.toURI());
                case "jar" -> {
                    final var pathElements = Paths.get(new URI(url.getFile())).toString().split("!");
                    if (pathElements.length == 0) {
                        yield null;
                    } else {
                        yield Paths.get(pathElements[0]);
                    }
                }
                default -> null;
            };
        } catch (final URISyntaxException e) {
            throw new RuntimeException(e);
        }
    }
}

File: src/test/java/utils/ProjectPaths.kt
package utils

import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import java.util.stream.Collectors
import kotlin.io.path.name
import kotlin.io.path.pathString
import kotlin.io.path.toPath

object ProjectPaths {

    val projectRoot: Path
        get() {
            var directory: Path? = projectRootByClass
            while (directory != null) {
                try {
                    val children = Files.list(directory).use {
                        it.map(Path::name).collect(Collectors.toSet())
                    }
                    if (children.contains("gradle") && children.contains("src")) {
                        return directory
                    }
                } catch (_: IOException) {
                }
                directory = directory.parent
            }
            error("Could not find project root")
        }

    private val projectRootByClass: Path
        get() {
            val replacedName = javaClass.name.replace(".", "/")
            val url = javaClass.getResource("/$replacedName.class")
            return when (url?.protocol) {
                "file" -> url.toURI().toPath()
                "jar" -> url.toURI().toPath().pathString.split("!").firstOrNull()?.let { Path.of(it) }
                else -> error("Unsupported protocol")
            } ?: error("Could not find project root")
        }
}
